### PR TITLE
Make `TestIntegrations/ReconcileLabels` a unit test

### LINKED ISSUE #31005: `TestIntegrations/ReconcileLabels` flakiness
#### Link(s) to logs

- https://github.com/gravitational/teleport/actions/runs/5969909774/job/16196600320

#### Relevant snippet

```
2023-08-24T23:37:31.9825302Z     integration_test.go:7923: 
2023-08-24T23:37:31.9825542Z         	Error Trace:	/__w/teleport/teleport/integration/integration_test.go:7923
2023-08-24T23:37:31.9825819Z         	            				/__w/teleport/teleport/integration/integration_test.go:125
2023-08-24T23:37:31.9826069Z         	Error:      	Timed out waiting for server update
2023-08-24T23:37:31.9826301Z         	Test:       	TestIntegrations/ReconcileLabels
```

### PR DESCRIPTION
**Pull request by atburke**
This PR replaces the flaky`TestIntegrations/ReconcileLabels` test with a less flaky unit test.

Fixes #31005.

### CODE REVIEW

**Comment by zmb3** [comment]
5 seconds seems pretty generous already - are we sure there's not something else we can do here?

**Comment by zmb3** [comment]
This still doesn't look correct. What exactly is it you're trying to do? Do you need a fake clock for this test or should we use a real one?

**Comment by zmb3** [comment]
After looking at this, I would test this differently. I think the behavior is pretty well isolated and could be replaced with a simpler unit test that's more reliable.

I'm thinking something like this:

- do test setup, create a server, add a node, etc
- spin up a goroutine that calls `ReconcileServerInfos` directly
- back in your main goroutine, call `clock.BlockUntil(1)`. This will wait for your reconciler goroutine to get to the first time.After. 
- When `BlockUntil` returns, you know the first reconcile loop completed, so you can now make your assertions that the labels and metadata are correct.
- Now cancel the context. No need to advance the clock. Canceling the context will cause the reconciler goroutine to complete. You can even send its error back to the main goroutine via channel to verify that it terminated correctly.

The only part that might be tricky is to ensure that your setup process creates an "unstarted" auth server. We don't want any of the background routines running, because `BlockUntil(1)` waits for exactly 1 sleeper, and we want to be sure that's the reconciler, not some other thing happening in the background.

This should be significantly faster and won't depend on time at all.

File: lib/auth/server_info_test.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	}, types.ServerInfoSpecV1{})
+	require.NoError(t, err)
+	serverInfo.SetSubKind(types.SubKindCloudInfo)
+	require.NoError(t, pack.a.UpsertServerInfo(ctx, serverInfo))
+
+	go pack.a.ReconcileServerInfos(ctx)
+	// Wait until the reconciler finishes processing the serverinfo.
+	clock.BlockUntil(1)
+	// Check that labels were received downstream.
+	require.Eventually(t, func() bool {
```
Comment (zmb3) [inline lib/auth/server_info_test.go:94]: You shouldn't need an eventually here. The `BlockUntil` call should guarantee that `setCloudLabelsOnNodes` has run by now.

I think you can replace this with a `require.Equal(t, labels, downstreamHandle.GetUpstreamLabels)`.

Unfortunately, there seems to be a separate issue. This test fails for me (with or without the eventually) 100% of the time with:

```
--- FAIL: TestReconcileLabels (0.20s)
    server_info_test.go:94: 
                Error Trace:    /Users/zmb/src/teleport/lib/auth/server_info_test.go:94
                Error:          Condition never satisfied
                Test:           TestReconcileLabels
FAIL
FAIL    github.com/gravitational/teleport/lib/auth      1.638s
```

**Review by zmb3 — COMMENTED** [review]
The code is looking better but the test is still failing.

File: lib/auth/server_info_test.go (hunk 1 above)
Comment (atburke) [reply lib/auth/server_info_test.go:100]: `setCloudNodesOnLabels` sends the labels asynchronously over the inventory control stream; they might not have been received by the time `setCloudNodesOnLabels` returns.

File: lib/auth/server_info_test.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+		return utils.StringMapsEqual(labels, downstreamHandle.GetUpstreamLabels(proto.LabelUpdateKind_SSHServerCloudLabels))
+	}, 100*time.Millisecond, 10*time.Millisecond)
```
Comment (zmb3) [inline lib/auth/server_info_test.go:102]: 100 milliseconds is rarely enough for CI. Let's try 3-5 seconds.

**Comment by zmb3** [comment]
Still fails - took 3 minutes this time.

```
teleport on atburke/reconcile-labels-flake [$] took 3s
➜ untilfail go test -run TestReconcileLabels ./lib/auth -count=10 -failfast
ok  	github.com/gravitational/teleport/lib/auth	7.055s
ok  	github.com/gravitational/teleport/lib/auth	7.154s
ok  	github.com/gravitational/teleport/lib/auth	6.972s
ok  	github.com/gravitational/teleport/lib/auth	7.007s
ok  	github.com/gravitational/teleport/lib/auth	6.992s
ok  	github.com/gravitational/teleport/lib/auth	6.988s
ok  	github.com/gravitational/teleport/lib/auth	7.019s
ok  	github.com/gravitational/teleport/lib/auth	7.101s
ok  	github.com/gravitational/teleport/lib/auth	7.014s
ok  	github.com/gravitational/teleport/lib/auth	7.092s
ok  	github.com/gravitational/teleport/lib/auth	7.139s
ok  	github.com/gravitational/teleport/lib/auth	6.990s
ok  	github.com/gravitational/teleport/lib/auth	7.092s
ok  	github.com/gravitational/teleport/lib/auth	7.037s
--- FAIL: TestReconcileLabels (0.28s)
    server_info_test.go:68:
        	Error Trace:	/Users/zmb/src/teleport/lib/auth/server_info_test.go:68
        	Error:      	Condition never satisfied
        	Test:       	TestReconcileLabels
FAIL
FAIL	github.com/gravitational/teleport/lib/auth	1.720s
FAIL

teleport on atburke/reconcile-labels-flake [$] took 2m58s
```

I wonder if there's a way we can mock out the control stream and just verify that `setCloudLabels` gets called with the right labels. That would eliminate all asynchronous logic and should be very reliable.

**Review by zmb3 — APPROVED** [review]
Now we're talking. The version with the mocked control stream is 3x faster and I've run it a couple thousand times without a failure.

**Review by lxea — APPROVED** [review]

**Pull request merged**